cmd/api/main.go

```diff
@@ -134,25 +126,45 @@ func main() {
 	hs.SetServingStatus("tekton.results.v1alpha2.Results", healthpb.HealthCheckResponse_SERVING)
 	healthpb.RegisterHealthServer(s, hs)
 
-	// Prometheus metrics
-	grpc_prometheus.Register(s)
+	// Start prometheus metrics server
+	prometheus.Register(s)
 	http.Handle("/metrics", promhttp.Handler())
 	go func() {
-		if err := http.ListenAndServe(":8080", promhttp.Handler()); err != nil {
+		log.Printf("Prometheus server listening on: %s", configFile.PROMETHEUS_PORT)
+		if err := http.ListenAndServe(":"+configFile.PROMETHEUS_PORT, promhttp.Handler()); err != nil {
 			log.Fatalf("error running Prometheus HTTP handler: %v", err)
 		}
 	}()
 
-	// Listen on port and serve.
-	port := os.Getenv("PORT")
-	if port == "" {
-		// Default gRPC server port to this value from tutorials (e.g., https://grpc.io/docs/guides/auth/#go)
-		port = "50051"
-	}
-	lis, err := net.Listen("tcp", ":"+port)
+	// Start gRPC server
+	lis, err := net.Listen("tcp", ":"+configFile.GRPC_PORT)
 	if err != nil {
 		log.Fatalf("failed to listen: %v", err)
 	}
-	log.Printf("Listening on :%s...", port)
-	log.Fatal(s.Serve(lis))
+	go func() {
+		log.Printf("gRPC server listening on: %s", configFile.GRPC_PORT)
+		log.Fatal(s.Serve(lis))
+	}()
+
+	// Load client TLS cert for client to connect to the gRPC server
+	creds, err = credentials.NewClientTLSFromFile(path.Join(configFile.TLS_PATH, "tls.crt"), configFile.TLS_OVERRIDE)
+	if err != nil {
+		log.Fatal("Error loading TLS certificate: ", err)
+	}
+
+	// Register gRPC server endpoint for gRPC gateway
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	mux := runtime.NewServeMux()
+	opts := []grpc.DialOption{grpc.WithTransportCredentials(creds)}
+	err = v1alpha2pb.RegisterResultsHandlerFromEndpoint(ctx, mux, ":"+configFile.GRPC_PORT, opts)
+	if err != nil {
+		log.Fatal("Error registering gRPC server endpoint: ", err)
+	}
+
+	// Start REST proxy server
+	log.Printf("REST server Listening on: %s", configFile.REST_PORT)
+	log.Fatal(http.ListenAndServeTLS(":"+configFile.REST_PORT, path.Join(configFile.TLS_PATH, "tls.crt"), path.Join(configFile.TLS_PATH, "tls.key"), mux))
```

Comment:
Not a bad idea that we serve TLS, but perhaps we only do this if the TLS cert and keys are present?

Configuring mTLS can be a bit of a pain - falling back to plain HTTP may help dev workflows or folks who want to use edge termination.

Reply:
Non TLS support added. If TLS files are not present, both gRPC and REST listner will start without TLS.